plugins/binanceExchange_ws.go

```diff
@@ -0,0 +1,283 @@
+
+	binance.WebsocketKeepalive = true
+
+	events := createStateEvents()
+
+	beWs := &binanceExchangeWs{
+		events:         events,
+		delimiter:      "",
+		assetConverter: model.CcxtAssetConverter,
+		mtx:            &sync.Mutex{},
+		streams:        make(map[string]*stream),
+	}
+
+	return beWs, nil
+}
+
+func getPrecision(floatStr string) int8 {
+
+	strs := strings.Split(floatStr, ".")
+
+	if len(strs) != 2 {
+		log.Printf("error get precision for float %s\n", floatStr)
+		return 0
+	}
+
+	return int8(len(strs[1]))
+}
+
+// GetTickerPrice impl.
+func (beWs *binanceExchangeWs) GetTickerPrice(pairs []model.TradingPair) (map[model.TradingPair]api.Ticker, error) {
+
+	priceResult := map[model.TradingPair]api.Ticker{}
+	for _, p := range pairs {
+
+		symbol, err := p.ToString(beWs.assetConverter, beWs.delimiter)
+
+		if err != nil {
+			return nil, err
+		}
+
+		tickerData, isTicker := beWs.events.SymbolStats.Get(symbol)
+
+		if !isTicker {
+			stream, err := subcribeTicker(symbol, beWs.events.SymbolStats)
+
+			if err != nil {
+				return nil, fmt.Errorf("error when subscribing for %s: %s", symbol, err)
+			}
+
+			//Store stream
+			beWs.mtx.Lock()
+			beWs.streams[symbol+STREAMTICKER] = stream
+			beWs.mtx.Unlock()
+
+			//Wait for binance to send events
+			time.Sleep(time.Second)
+
+			tickerData, isTicker = beWs.events.SymbolStats.Get(symbol)
+
+			//We couldn't subscribe for this pair
+			if !isTicker {
+				return nil, fmt.Errorf("error while fetching ticker price for trading pair %s", symbol)
+			}
+
+		}
+
+		//Show how old is the ticker
+		log.Printf("Ticker for %s is %d milliseconds old!\n", symbol, time.Now().Sub(tickerData.createdAt).Milliseconds())
+
+		if isStale(tickerData, TTLTIME) {
+			return nil, fmt.Errorf("ticker for %s symbols is older than %v", symbol, TTLTIME)
+		}
+
+		tickerI := tickerData.data
+
+		//Convert to WsMarketStatEvent
+		ticker, isOk := tickerI.(*binance.WsMarketStatEvent)
+
+		if !isOk {
+			return nil, ErrConversionWsMarketEvent
+		}
+
+		askPrice, e := strconv.ParseFloat(ticker.AskPrice, 64)
+		if e != nil {
+			return nil, fmt.Errorf("unable to correctly parse 'ask': %s", e)
+		}
+		bidPrice, e := strconv.ParseFloat(ticker.BidPrice, 64)
+		if e != nil {
+			return nil, fmt.Errorf("unable to correctly parse 'bid': %s", e)
+		}
+		lastPrice, e := strconv.ParseFloat(ticker.LastPrice, 64)
+		if e != nil {
+			return nil, fmt.Errorf("unable to correctly parse 'last': %s", e)
+		}
+
+		priceResult[p] = api.Ticker{
+			AskPrice:  model.NumberFromFloat(askPrice, getPrecision(ticker.AskPrice)),
+			BidPrice:  model.NumberFromFloat(bidPrice, getPrecision(ticker.BidPrice)),
+			LastPrice: model.NumberFromFloat(lastPrice, getPrecision(ticker.LastPrice)),
+		}
+	}
+
+	return priceResult, nil
+}
+
+//Unsubscribe ... unsubscribe from binance streams
+func (beWs *binanceExchangeWs) Unsubscribe(stream string) {
+
+	beWs.mtx.Lock()
+
+	if stream, isStream := beWs.streams[stream]; isStream {
+		stream.Close()
+	}
```

Comment:
do we also need to remove all values from `beWs.streams`?

Reply:
We can remove them, but it's only if we want this

Reply:
yes lets' remove please